cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/input/BasicPartitionedFileSetInputContext.java

```diff
@@ -42,34 +49,74 @@
 
   private static final Type STRING_PARTITION_KEY_MAP_TYPE = new TypeToken<Map<String, PartitionKey>>() { }.getType();
 
-  private final Path inputPath;
-  private final String mappingString;
+  private final Map<String, PartitionKey> pathToPartitionMapping;
 
-  private PartitionKey partitionKey;
+  private final boolean isCombineInputFormat;
+  private final Configuration conf;
+
+  private final List<Path> inputPaths;
+  private List<PartitionKey> partitionKeys;
+
+  // for caching in case of CombineFileInputFormat
+  private String currentInputfileName;
+  private PartitionKey currentPartitionKey;
 
   BasicPartitionedFileSetInputContext(MultiInputTaggedSplit multiInputTaggedSplit) {
     super(multiInputTaggedSplit.getName());
 
     InputSplit inputSplit = multiInputTaggedSplit.getInputSplit();
-    if (!(inputSplit instanceof FileSplit)) {
-      throw new IllegalArgumentException(String.format("Expected a '%s', but got '%s'.",
-                                                       FileSplit.class.getName(), inputSplit.getClass().getName()));
+    inputPaths = new ArrayList<>();
+    if (inputSplit instanceof FileSplit) {
+      isCombineInputFormat = false;
+      inputPaths.add(((FileSplit) inputSplit).getPath());
+    } else if (inputSplit instanceof CombineFileSplit) {
+      isCombineInputFormat = true;
+      Collections.addAll(inputPaths, ((CombineFileSplit) inputSplit).getPaths());
+    } else {
+      throw new IllegalArgumentException(String.format("Expected either a '%s' or a '%s', but got '%s'.",
+                                                       FileSplit.class.getName(), CombineFileSplit.class.getName(),
+                                                       inputSplit.getClass().getName()));
     }
-    this.inputPath = ((FileSplit) inputSplit).getPath();
-    this.mappingString = multiInputTaggedSplit.getConf().get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+
+    this.conf = multiInputTaggedSplit.getConf();
+    String mappingString = conf.get(PartitionedFileSetDataset.PATH_TO_PARTITIONING_MAPPING);
+    this.pathToPartitionMapping =
```

Comment:
It will be approximately 15KB for every 1000 splits.
So 50k splits will be 7.5MB.

(Approximated 120 bytes for the String key and 30 bytes for the PartitionKey object, for each entry in the map).